File: js/BeginnersSelection/04_Coins.js
// https://atcoder.jp/contests/abs/tasks/abc087_b

Object.prototype.let = function(f) {
  return f(this.valueOf());
}

require('fs')
  .readFileSync('/dev/stdin', 'utf8')
  .split('\n')
  .map(x => +x)
  .let(([a, b, c, x]) =>
    [...Array(Math.min(a, Math.floor(x / 500)) + 1).keys()]
      .map(i =>
        [...Array(Math.min(b, Math.floor((x - 500 * i) / 100)) + 1).keys()]
          .filter(j => x - 500 * i - 100 * j <= 50 * c)
          .length
      )
  )
  .reduce((a, x) => a + x)
  .let(console.log);

// https://atcoder.jp/contests/abs/submissions/21478036

File: py/BeginnersSelection/04_Coins.py
# https://atcoder.jp/contests/abs/tasks/abc087_b

from math import floor

class W:
  def __init__(self, v): self.v = v
  let = lambda self, f: W(f(self.v))
  map = lambda self, f: W(map(f, self.v))
  filter = lambda self, f: W(filter(f, self.v))

W((input(), input(), input(), input())) \
  .map(int) \
  .let(tuple) \
  .let(lambda t:
    W(range(0, min([t[0], floor(t[3] / 500)]) + 1)) \
      .map(lambda i:
        W(range(0, min([t[1], floor((t[3] - 500 * i) / 100)]) + 1)) \
          .filter(lambda j: t[3] - 500 * i - 100 * j <= 50 * t[2]) \
          .let(list)
          .let(len) \
          .v
      ) \
      .let(sum) \
      .v
  ) \
  .let(print)

# https://atcoder.jp/contests/abs/submissions/21624786
